mergify_engine/squash_pull.py

```diff
@@ -0,0 +1,119 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import typing
+from mergify_engine import github_types
+from mergify_engine import context
+from mergify_engine.clients import http
+import dataclasses
+
+
+@dataclasses.dataclass
+class SquashFailure(Exception):
+    reason: str
+
+
+async def squash_pull(ctxt: context.Context) -> None:
+
+    if ctxt.pull["commits"] <= 1:
+        return
+
+    pull_number = ctxt.pull["number"]
+    head_branch = ctxt.pull["head"]["ref"]
+    head_owner = ctxt.pull["head"]["user"]["login"]
+    head_repo = ctxt.pull["head"]["repo"]["name"]
+    base_owner = ctxt.pull["base"]["user"]["login"]
+    base_repo = ctxt.pull["base"]["repo"]["name"]
+
+    try:
+        # get the branch
+        branch = typing.cast(
+            github_types.GitHubBranchRef,
+            await ctxt.client.item(
+                f"/repos/{head_owner}/{head_repo}/git/ref/heads/{head_branch}"
+            ),
+        )
+    except http.HTTPClientSideError:  # pragma: no cover
+        raise ValueError("ERROR: get the branch")
+
+    try:
+        # get the last commit
+        sha = typing.cast(github_types.SHAType, branch["object"]["sha"])
+        commit = typing.cast(
+            github_types.GitHubCommit,
+            await ctxt.client.item(
+                f"/repos/{head_owner}/{head_repo}/git/commits/{sha}"
+            ),
+        )
+    except http.HTTPClientSideError:  # pragma: no cover
+        raise ValueError("ERROR: get the last commit")
+
+    try:
+        # get commits' pull to have its first commit
+        commits = [
+            commits
+            async for commits in typing.cast(
+                typing.AsyncGenerator[github_types.GitHubCommit, None],
+                ctxt.client.items(f"{ctxt.base_url}/pulls/{pull_number}/commits"),
+            )
+        ]
+    except http.HTTPClientSideError:  # pragma: no cover
+        raise ValueError("ERROR: get commits' pull to have its first commit")
+
+    try:
+        # create commit
+        response = await ctxt.client.post(
+            f"/repos/{base_owner}/{base_repo}/git/commits",
+            json={
+                "owner": head_owner,
+                "repo": head_repo,
+                "message": commits[0]["commit"]["message"],
+                "tree": commit["tree"]["sha"],
+                "parents": [ctxt.pull["base"]["sha"]],
+            },
+        )
+    except http.HTTPClientSideError:  # pragma: no cover
+        raise ValueError("ERROR: create commit")
+
+    try:
+        # create ref with the new commit
+        new_commit = typing.cast(github_types.GitHubCommit, response.json())
+        await ctxt.client.post(
+            f"/repos/{base_owner}/{base_repo}/git/refs",
+            json={
+                "owner": base_owner,
+                "repo": base_repo,
+                "ref": f"refs/heads/{head_branch}",
```

Comment:
I don't think we need to create this ref, just updating the head_owner/head_repo one should be enough